domino/flyte/task.py

```diff
@@ -0,0 +1,300 @@
+        if not self.EnvironmentRevisionSpec: unresolved_fields.append("environmentRevisionSpec")
+        if self.ComputeClusterProperties and not self.ComputeClusterProperties.is_resolved(): unresolved_fields.append("computeClusterProperties")
+        if not self.VolumeSizeGiB: unresolved_fields.append("volumeSizeGiB")
+        if not self.DatasetSnapshots: unresolved_fields.append("datasetSnapshots")
+        if self.ExternalVolumeMountIds == None: unresolved_fields.append("externalVolumeMountIds")
+
+        return unresolved_fields
+    
+
+    def resolve_job_properties(self):
+        if self.is_resolved():
+            click.secho("Job properties are already fully resolved")
+            return
+        
+        click.secho("Retrieving default properties for job")
+        # TODO: Can we make this work outside of runs? Also can we modify this so we don't need owner/project in the config?
+        url = f"{os.environ['DOMINO_API_PROXY']}/v4/jobs/{self.OwnerName}/{self.ProjectName}/resolveJobDefaults"
+        payload = self.to_json()
+
+        response = requests.post(url, json=payload)
+        if response.status_code != 200:
+            raise Exception(f"Failed to resolve job properties (StatusCode {response.status_code}): {response.text})")
+        
+        resolved_job_config = response.json()
+
+        self.CommitId = resolved_job_config["commitId"]
+        self.MainRepoGitRef = GitRef.from_json(resolved_job_config["mainRepoGitRef"]) if resolved_job_config.get("mainRepoGitRef") else None
+        self.HardwareTierId = resolved_job_config["hardwareTierId"]["value"]
+        self.EnvironmentId = resolved_job_config["environmentId"]
+        self.EnvironmentRevisionSpec = EnvironmentRevisionSpecification.from_json(resolved_job_config["environmentRevisionSpec"])
+        self.ComputeClusterProperties = ClusterProperties.from_json(resolved_job_config["computeClusterProperties"]) if resolved_job_config.get("computeClusterProperties") else None
+        self.VolumeSizeGiB = resolved_job_config["volumeSizeGiB"]
+        self.DatasetSnapshots = [DatasetSnapshot.from_json(snapshot_json) for snapshot_json in resolved_job_config["datasetSnapshots"]]
+        self.ExternalVolumeMountIds = resolved_job_config["externalVolumeMountIds"]
+
+        click.secho(f"Resolved job properties: {self}", fg="cyan")
+
+
+    def to_json(self) -> Dict[str, Any]:
+        return {
+            "ownerName": self.OwnerName,
+            "projectName": self.ProjectName,
+            "apiKey": self.ApiKey,
+            "command": self.Command,
+            "commitId": self.CommitId,
+            "hardwareTierId": { "value": self.HardwareTierId } if self.HardwareTierId else None,
+            "environmentId": self.EnvironmentId,
+            "environmentRevisionSpec": self.EnvironmentRevisionSpec.to_json() if self.EnvironmentRevisionSpec else None,
+            "datasetSnapshots": [snapshot.to_json() for snapshot in self.DatasetSnapshots],
```

Comment:
```suggestion
            "datasetSnapshots": [snapshot.to_json() for snapshot in self.DatasetSnapshots] if self.DatasetSnapshots else [],
```
Also, an empty list in Python is "false-y" which may or may not affect the various `is_resolved` logic in this class.